proof/crefine/RISCV64/Schedule_C.thy

```diff
@@ -1543,12 +1535,336 @@ lemma refill_budget_check_ccorres:
+                                and fa="\<lambda>r. rAmount_update (\<lambda>_. rAmount r + ksConsumedTime s) r"
+                                 in rf_sr_refill_update2; fastforce?)
+                       apply (fastforce simp: sched_context.expand)
+                      apply (clarsimp simp: typ_heap_simps)
+                      apply (fastforce simp: typ_heap_simps' sc_ptr_to_crefill_ptr_def
+                                             crefill_relation_def)
+                     apply (fastforce simp: csched_context_relation_def)
+                    apply (fastforce dest: crefill_relationD rf_sr_ksConsumed
+                                     simp: typ_heap_simps' sc_ptr_to_crefill_ptr_def
+                                           crefill_relation_def)
+                   apply (rule_tac Q'="\<lambda>_ s. ?abs s \<and> active_sc_at' scPtr s \<and> scPtr = ksCurSc s
+                                             \<and> scPtr = scPtr' \<and> consumed = ksConsumedTime s"
+                                in hoare_post_imp)
+                    apply (clarsimp simp: active_sc_at'_def)
+                    apply normalise_obj_at'
+                    apply (frule (1) sc_ko_at_valid_objs_valid_sc'[OF _ invs_valid_objs'])
+                    apply (clarsimp simp: valid_sched_context'_def)
+                   apply (wpsimp wp: updateRefillHd_invs')
+                  apply (clarsimp simp: guard_is_UNIV_def)
+                 apply (ctac add: refill_budget_check_ccorres)
+                apply (wpsimp wp: isRoundRobin_wp)
+               apply (vcg exspec=isRoundRobin_modifies)
+              apply (fastforce intro: ccorres_return_Skip)
+             apply ceqv
+            apply (rule_tac P'="\<lambda>s. sc_at' scPtr s \<and> scPtr = ksCurSc s \<and> consumed = ksConsumedTime s"
+                         in updateSchedContext_ccorres_lemma3[where P=\<top>])
+              apply vcg
+             apply simp
+            subgoal
+              by (fastforce dest: rf_sr_ksCurSC rf_sr_ksConsumed
+                           intro: rf_sr_sc_update_no_refill_buffer_update2
+                                  refill_buffer_relation_sc_no_refills_update
+                            simp: typ_heap_simps' csched_context_relation_def option_to_ctcb_ptr_def)
+           apply (wpsimp wp: isRoundRobin_wp)
+          apply (clarsimp simp: guard_is_UNIV_def)
+         apply (fastforce intro: ccorres_return_Skip)
+        apply ceqv
+       apply (rule ccorres_from_vcg[where P=\<top> and P'=UNIV])
+       apply (rule allI, rule conseqPre, vcg)
+       apply (clarsimp simp: setConsumedTime_def rf_sr_def cstate_relation_def Let_def
+                             modify_def get_def put_def bind_def carch_state_relation_def
+                             cmachine_state_relation_def)
+      apply wpsimp
+     apply (clarsimp simp: guard_is_UNIV_def)
+    apply (wpsimp wp: scActive_wp)
+   apply (vcg exspec=sc_active_modifies)
+  apply clarsimp
+  apply safe
```

Comment:
Despite its name `safe` is not such a safe tactic :-). Would it make sense to say `apply (rule conjI; clarsimp)` instead here? Or maybe even `apply (intro conjI impI)`. That would provide some check against the goal form that is expected here and also for the subgoals that are being produced. With `safe` after a change you might get more or fewer goals and it'll be hard to match up the following proofs to what was expected.

Reply:
Just `apply (rule conjI; clarsimp)` won't quite do it, but I can just replace the `safe` with `apply (intro conjI impI allI; clarsimp)` and it'll work with the rest of the proof unchanged.

I sometimes use `safe` when I'm exploring and not sure whether I've got provable subgoals. Is it a reasonable thing to use then? Or is it better to just go with the good old `intro conjI impI allI`?

Reply:
For exploration it's fine, because you will see if it misbehaves, but I would leave something more controlled in the final proof, so I usually just do the `intro conjI impI allI`. Come to think of it, it is so common that we might want to give it a name so that it is as little typing as `safe`, and actually safe in the sense we want.